### DISCUSSION TITLE
Helper for animated page transitions

### DISCUSSION THREAD

**Discussion by bravokiloecho** [vercel/next.js#32218] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

I would like the ability to animate between pages, with either CSS transitions or with hooks for JS to use.

## Describe the solution you'd like

Basically what nuxt does: https://nuxtjs.org/api/pages-transition/

## Describe alternatives you've considered

There is third party solution (https://www.npmjs.com/package/next-page-transitions) but this only works with CSS and risks falling behind any changes to the next.js core.

## Additional context

I feel like animated page transitions is a basic feature of modern web development and deserves to be included in the core.

**Comment by kasim444** [comment]
You can easily make page transition using framer-motion. The example below is the method I use, you can customize it as you like. If you wrap the component you created under pages with Layout, you will get what you want. I hope it works for you.


Wrapper -> Layout.ts
```
import * as React from 'react';
import Head from 'next/head';
import { Header, Footer } from '../organisms';
import { motion } from 'framer-motion';

export type LayoutProps = {
  title: string;
};

const duration = 0.35;

const variants = {
  initial: {
    opacity: 0,
  },
  enter: {
    opacity: 1,
    transition: {
      duration: duration,
      delay: duration,
      when: 'beforeChildren',
    },
  },
  exit: {
    opacity: 0,
    transition: { duration: duration },
  },
};

const Layout: React.FunctionComponent<LayoutProps> = ({
  children,
  title,
}) => {

  return (
    <motion.div
      variants={variants}
      initial="initial"
      animate="enter"
      exit="exit"
    >
      <Head>
        <title>{title}</title>
      </Head>
      <Header/>
      {children}
      <Footer />
    </motion.div>
  );
};

export default Layout;
```

Homepage -> Whatever.ts
```
import Layout from '../components/layout/layout';

const IndexPage = () => {
  return (
    <Layout title="Home page">
      <h1>Hello from home page 🖖</h1>
    </Layout>
  );
};
export default IndexPage;
```

**Comment by bravokiloecho** [comment]
Hey @kasim444, thanks for the tip. This does look good, but ideally page transitions shouldn't be tied to a particular animation library.

There are lots of good ways to animate in react and I think a developer should be able to animate page transitions without committing to a particular library.

**Comment by trompx** [comment]
Hey @timneutkens I was wondering if you were still considering that feature?

I have many animations/transitions in gsap in legacy projects and it would be best to have some hook/events we can trigger manually so the system could be used with multiple libraries. Most of the transition packages are still limited in that regard (either just allow CSS transitions, force to declare transition duration upfront, can't handle entering the new page and exiting the previous one at the same time: motion from framer has a beta feature but as @bravokiloecho mentionned it would tie us to that library which is not optimal).

We would then easily be able to setup OutThenIn / InThenOut / OutAndIn (sometimes you want to wait that the next page is loaded to start animating both pages at the same time) kind of transitions.
